### DISCUSSION TITLE
Can language be set without setting script

### DISCUSSION THREAD

**Discussion by murasu** [harfbuzz/harfbuzz#2877] [Q&A] [answered]
I'm writing a tool that visually compares the shaping outcome of a string through CoreText and harfbuzz. In this case, I'm using NotoSansDevanagari. When I call the shapers without setting the script or language, the outcomes match - as expected.

To test the same string in a non-default language, in this case Marathi, I tried the following:

CoreText: I set `kCTLanguageAttributeName` in the attributed string and CoreText gives me the correct glyph for Marathi LA. There is no need for me to specify that the script is Devanagari.

harfbuzz: I need to set the language as `"mr"`  _and_  `HB_SCRIPT_DEVANAGARI` as the script get the same results. 

```
hb_buffer_set_direction(buf, HB_DIRECTION_LTR);
hb_buffer_set_script(buf, HB_SCRIPT_DEVANAGARI);
hb_buffer_set_language(buf, hb_language_from_string("mr", -1));
```

Omitting the script or setting it to `HB_SCRIPT_UNKNOWN` does not help (green is hb):

<img width="859" alt="Notification_Center" src="https://user-images.githubusercontent.com/22926756/108827758-f193b980-7600-11eb-9890-358311937be1.png">

I'm wondering if it's possible to call hb_shape without specifically setting the script when the string needs to be shaped for a non-default language. If it's not, should I seek the `help of hb_unicode_script()` to first determine the script and set it?

What do the applications that use hb for shaping generally do?

**Comment by murasu** [comment]
Answering my own question:

`hb_buffer_guess_segment_properties(buf);`

Sets the script.

**Comment by behdad** [comment, marked as answer]
The OpenType shaper needs the script.
